presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -222,12 +297,61 @@ private PlanNodeStatsEstimate estimateLogicalOr(Expression left, Expression righ
                 return PlanNodeStatsEstimate.unknown();
             }
 
-            PlanNodeStatsEstimate logicalAndEstimate = new FilterExpressionStatsCalculatingVisitor(leftEstimate, session, types).process(right);
-            if (logicalAndEstimate.isOutputRowCountUnknown()) {
+            PlanNodeStatsEstimate andEstimate = new FilterExpressionStatsCalculatingVisitor(leftEstimate, session, types).process(right);
+            if (andEstimate.isOutputRowCountUnknown()) {
                 return PlanNodeStatsEstimate.unknown();
             }
-            PlanNodeStatsEstimate sumEstimate = addStatsAndSumDistinctValues(leftEstimate, rightEstimate);
-            return differenceInNonRangeStats(sumEstimate, logicalAndEstimate);
+
+            PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
+            double outputRowCount = min(leftEstimate.getOutputRowCount() + rightEstimate.getOutputRowCount() - andEstimate.getOutputRowCount(), input.getOutputRowCount());
+            verify(outputRowCount >= 0, "outputRowCount must be greater than or equal to zero: %s", outputRowCount);
+            result.setOutputRowCount(outputRowCount);
+
+            if (outputRowCount == 0) {
+                return process(FALSE_LITERAL);
+            }
+
+            input.getSymbolsWithKnownStatistics().forEach(symbol -> {
```

Comment:
What if we have methods as:

```suggestion
/**
 * Subtracts stats that are assumed to be subset of left stats
 */
PlanNodeStatsEstimate subtractSubset(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right) {
  PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
  double outputRowCount = left.getOutputRowCount() - right.getOutputRowCount();
  verify(outputRowCount >= 0, "outputRowCount must be greater than or equal to zero: %s", outputRowCount);
  result.setOutputRowCount(outputRowCount);

  left.getSymbolsWithKnownStatistics().forEach(symbol -> {
    SymbolStatsEstimate leftSymbolStats = left.getSymbolStatistics(symbol);
    SymbolStatsEstimate rightSymbolStats = right.getSymbolStatistics(symbol);
    
    SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
    
    newSymbolStats.setLowValue(minNonNan(leftSymbolStats.getLowValue(), rightSymbolStats.getLowValue()));
    newSymbolStats.setHighValue(maxNonNan(leftSymbolStats.getHighValue(), rightSymbolStats.getHighValue()));
    
    double newDistinctValues = leftSymbolStats.getDistinctValuesCount() - rightSymbolStats.getDistinctValuesCount();
    
    double leftNumberOfNulls = leftSymbolStats.getNullsFraction() * leftEstimate.getOutputRowCount();
    double rightNumberOfNulls = rightSymbolStats.getNullsFraction() * rightEstimate.getOutputRowCount();
     
    double newNumberOfNulls = leftNumberOfNulls - rightNumberOfNulls;
    double newNullsFraction = min(newNumberOfNulls, outputRowCount) / outputRowCount;
     
    newSymbolStats.setNullsFraction(newNullsFraction);
    result.addSymbolStatistics(symbol, newSymbolStats.build());
  }
  
  return result.build();
}

PlanNodeStatsEstimate cap(PlanNodeStatsEstimate input, PlanNodeStatsEstimate cap) {
  PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
  double outputRowCount = min(input.getOutputRowCount(), cap.getOutputRowCount());
  result.setOutputRowCount(outputRowCount);
        
  input.getSymbolsWithKnownStatistics().forEach(symbol -> {
    SymbolStatsEstimate inputSymbolStats = input.getSymbolStatistics(symbol);
    SymbolStatsEstimate capSymbolStats = cap.getSymbolStatistics(symbol);
    
    SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
    
    // TODO: this should be capped too
    newSymbolStats.setLowValue(inputSymbolStats.getLowValue());
    newSymbolStats.setHighValue(inputSymbolStats.getHighValue());
    
    double newDistinctValues = minNonNan(inputSymbolStats.getDistinctValuesCount() , capSymbolStats.getDistinctValuesCount());
    
    double inputNumberOfNulls = inputSymbolStats.getNullsFraction() * inputEstimate.getOutputRowCount();
    double capNumberOfNulls = capSymbolStats.getNullsFraction() * capEstimate.getOutputRowCount();
     
    double newNumberOfNulls = minNonNan(inputNumberOfNulls, capNumberOfNulls);
    double newNullsFraction = min(newNumberOfNulls, outputRowCount) / outputRowCount;
     
    newSymbolStats.setNullsFraction(newNullsFraction);
    result.addSymbolStatistics(symbol, newSymbolStats.build());
  }
  
  return result.build();
}
```

Then for `OR` you simply do:
```
cap(subtractSubset(addStats(leftStats, rightStats), andStats), inputStats);
```

I think its more modular and easier to grasp. What do you think?